### PR TITLE
Stabilize the useResourcePermissions hook

### PR DESCRIPTION
**Pull request by adamziel**
## What?
It's been two months since the `__experimentalUseResourcePermissions` hook was introduced in PR https://github.com/WordPress/gutenberg/pull/38785. This PR stabilizes it.

CC @gziolo @Mamaduka @kevin940726

### CODE REVIEW

**Comment by gziolo** [comment]
Thank you for going back to make this API stable. I support the idea!

 Do you know why `useResourcePermission` is used only in one place? Isn't it something that would also be useful to integrate with the Reusable block or the Template Part block?

I would expect to see the JSDoc for the stable now function included in the README file of the `@wordpress/core-data` package. As part of the documentation process, we should also include it in the CHANGELOG file of the package.

As for the dev note, it's something that should be compiled together with other changes to the package – new API for `useEntityRecord` and `useEntityRecords`.

**Comment by adamziel** [comment]
@gziolo I just added the README and Changelog entries, good call!

> Do you know why useResourcePermission is used only in one place? Isn't it something that would also be useful to integrate with the Reusable block or the Template Part block?

Good call as well, I can see other places where this would be useful. Let's pursue the refactor separately from this PR.

File: packages/core-data/src/index.js [hunk 1]
```diff
@@ -70,7 +70,7 @@ register( store );
 export { default as EntityProvider } from './entity-provider';
 export { default as useEntityRecord } from './hooks/use-entity-record';
 export { default as useEntityRecords } from './hooks/use-entity-records';
-export { default as __experimentalUseResourcePermissions } from './hooks/use-resource-permissions';
+export { default as useResourcePermissions } from './hooks/use-resource-permissions';
```
Comment (gziolo) [inline packages/core-data/src/index.js]: It looks like a good opportunity to re-export those APIs directly from `./hooks` that is used to generate API docs.

**Review by gziolo — APPROVED** [review]
Excellent, looks good from my perspective after the revisions were applied.

> Good call as well, I can see other places where this would be useful. Let's pursue the refactor separately from this PR

Awesome, it would be great to tackle it in parallel to confirm that the stable API is flexible enough.

File: packages/core-data/src/index.js [hunk 2]
```diff
@@ -68,9 +68,6 @@ export const store = createReduxStore( STORE_NAME, storeConfig() );
 register( store );
 
 export { default as EntityProvider } from './entity-provider';
-export { default as useEntityRecord } from './hooks/use-entity-record';
```
Comment (gziolo) [inline packages/core-data/src/index.js:71]: Was it exported twice? 😅

Comment (adamziel) [reply packages/core-data/src/index.js:71]: yes 😅

**Comment by adamziel** [comment]
Here's the PR where I explore the refactor: https://github.com/WordPress/gutenberg/pull/43305

**Comment by gziolo** [comment]
We have now a few more places to update `__experimentalUseResourcePermissions` and we are done 🎉

File: packages/core-data/README.md [hunk 3]
```diff
@@ -875,6 +875,47 @@ _Returns_
 
 -   `EntityRecordsResolution< RecordType >`: Entity records data.
 
+### useResourcePermissions
+
+Resolves resource permissions.
```
Comment (gziolo) [inline packages/core-data/README.md:880]: The example is helpful, but we could include more details about what type of permissions it covers, so people have a better overview when scanning the docs.

**Comment by gziolo** [comment]
In other news, this is the last task that concludes the proposal from https://make.wordpress.org/core/2022/07/04/proposal-better-rest-api-handling-in-javascript/. The only task left is compiling a dev note covering all final changes and enriching the documentation with the new APIs after WordPress 6.1 is out. Everything here is built on top of existing public APIs, so that means we eventually need to decrease the visibility of low-level APIs in the Block Editor Handbook.

Awesome work @adamziel 🚀

**Review by gziolo — APPROVED** [review]

File: packages/core-data/README.md (hunk 3 above)
Comment (adamziel) [reply packages/core-data/README.md:880]: I added a second example, thank you!

**Comment by Mamaduka** [comment]
Thanks, @adamziel. I agree that the hook was ready for stabilization.

The one issue I would like to see resolved, and it's not entirely related to this hook, is the one I mentioned in https://github.com/WordPress/gutenberg/pull/35527#issuecomment-1184202416.

When using this hook, even if you only check for single permission, it will still make multiple calls to the same endpoint.

While a few extra API calls aren't necessarily bad, numbers will add up as we continue adoption in the core.

## Example
```js
var el = wp.element.createElement;
var __ = wp.i18n.__;
var registerPlugin = wp.plugins.registerPlugin;
var PluginDocumentSettingPanel = wp.editPost.PluginDocumentSettingPanel;
var useResourcePermissions = wp.coreData.useResourcePermissions;
 
function MyDocumentSettingPlugin() {
    // Replace tag ID.
    const { canUpdate } = useResourcePermissions( 'tags', 20 );

    if ( ! canUpdate ) {
        return null;
    }

    return el(
        PluginDocumentSettingPanel,
        {
            className: 'my-document-setting-plugin',
            title: 'Permissions',
        },
        __( 'User has required Permissions' )
    );
}
 
registerPlugin( 'my-document-setting-plugin', {
    render: MyDocumentSettingPlugin,
} );
```

**Comment by adamziel** [comment]
@Mamaduka good note! As you noticed, it is not specific to this hook - see this [patch for `canUser` resolver](https://github.com/WordPress/gutenberg/pull/43480) I just prepared.

**Comment by adamziel** [comment]
Dev note proposed at https://make.wordpress.org/core/?p=99829&preview=1&_ppp=2b1db975ea

**Pull request merged**